### ISSUE TITLE
NT path is not available in SSH

### ISSUE THREAD

**Issue by poma** [Microsoft/WSL#2145]
I currently run a background ssh server (dropbear) at Windows startup and to use Bash I use Putty that connects to localhost. In my case NT Path variable is not imported in Ubuntu. Looks like currently Path is somehow hard-injected at wsl startup and is not very flexible.

Maybe better approach would be to somehow expose Windows variables (/proc, syscall, whatever) and then have a small user script that appends path to current environment at startup so that users can customize how and when it is appended?

**Comment by benhillis** [comment]
@poma - Good suggestion, an easy way for users to convert NT->LX and LX->NT paths has come up a few times.  Your idea of a virtual file is an interesting idea.

**Comment by therealkenc** [comment]
>Your idea of a virtual file is an interesting idea.

You could expose the registry entries in `/proc`, sure...
```
HKEY_CURRENT_USER\Environment
Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Environment
```
That _would_ be interesting idea. [Heck, it would be "interesting" if the whole registry tree were a read/write filesystem under `/proc` or elsewhere. But I digress.]

But using the registry for WSL's definition of `$PATH` is kind of the wrong layer. Here is our current behavior:
```
C:\>set PATH=/foo
C:\>echo %PATH%
/foo
C:\>C:\Windows\System32\bash.exe
ken@DESKTOP-9BDQA3V:/mnt/c$ echo $PATH
/usr/local/sbin/:/usr/local/bin: ...blah blah.... /mnt/c/Windows/System32:/mnt/c/Windows:/mnt/c/Windows/System32/wbem:/mnt/c/Windows/System32/WindowsPowerShell/v1.0:/mnt/c/Program Files/Microsoft SQL .....
```
_"That ain't right"_. Nor is `$PATH` a special case here. You could make the same argument for `$TERM` (which came up recently), `$HOMEPATH`, etc. Or for that matter `$INCLUDE`, `$WindowsSdkDir`, `$WindowsLibPath` (ad nausium). That is, if one is making the case that you should be able to launch `cl.exe` using WSL introp. Which, at least on paper, we are, right? The `$PATH` to `cl.exe` doesn't even live in the registry, because, you can't have too many versions of the thing in your filesystem.

**Comment by therealkenc** [comment]
[`WSLENV`](https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-17063) was added in [17063](https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-17063).

**Comment by ketozhang** [comment]
I don't understand how the addition of WSLENV fix this problem.

Do I have to add some sort of injection of `$env:path` to `WSLPATH=$env:path/u` then add that to WSL $PATH later (which doesnt work)?

**Comment by DrVanScott** [comment]
From my point of view, WSLENV doesn't help here as WSLENV is itself not set if the shell is started via ssh.

A WSL login shell shouldn't behave differently for
* direct invocation
* ssh login

e.g

    echo $WSLENV

or

    cmd.exe

should behave the same

**Comment by therealkenc** [comment]
> I don't understand how the addition of WSLENV fix this problem.

You're right, actually. It doesn't. This issue went sideways on OP second paragraph "Maybe better approach would be to somehow expose Windows variables". The approach taken was `$WSLENV` (for better or worse), and this submission went down with it.

The _title_ and first paragraph of the OP, meanwhile, was about `ssh`/`sshd` behavior. Which is not a WSL topic at all, ref #4911.

**Comment by ketozhang** [comment]
> The title and first paragraph of the OP, meanwhile, was about ssh/sshd behavior. Which is not a WSL topic at all, ref #4911.

Getting back to the topic of the title _"NT path is not available in SSH"_, in #4911 the example was

```
> ssh user@host
$ /mnt/c/Windows/System32/cmd.exe /C echo test
```

We definitely can access to any NT command using full path of course. The behavior I'd _expect_ would be to use all environment as if `$env:path` was injected as it is when using `wsl` hence being able to shorten the command to just `cmd test`.  However, this isn't true as you said SSH/SSHD doesn't do this.

Do you have any suggestion on repeating this injection. Maybe it should be a command in`~/.profile, ~/.bashrc, ...` for those that want the behavior.

**Comment by therealkenc** [comment]
> The behavior I'd expect would be to use all environment as if $env:path was injected as it is when using wsl

That would be an unusual thing to expect, since the `ssh.exe` application has never heard of WSL, and the `sshd` app on the Linux side doesn't even know it is running in WSL. 

WSL, for its part, knows nothing of either `ssh` or `sshd`.

> Do you have any suggestion on repeating this injection

Yes, in the first reply of #4911 which is still open (or at least, not quite dead yet).

Contrast this closed issue, which was a request to quoth: "somehow expose Windows variables (/proc, syscall, whatever)". Which for better or for worse, addressed with `$WSLENV` but does not help the OP's use case.
